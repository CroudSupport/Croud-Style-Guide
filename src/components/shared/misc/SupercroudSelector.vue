<template>
    <div class="dropdown-wrapper">
        <semantic-form-dropdown
            ref="supercroudSelector"
            v-model="model"
            :settings="dropdownSettings"
            :options="filteredOptions"
            :data-tooltip="summary"
            :placeholder="dropdownPlaceholder"
            :data-inverted="true"
            :auto-update="false"
            multiple
            class="team-aware">

            <div v-for="(users, team) in teams" class="header item" :data-value="team">
                <i class="dropdown icon"></i>
                <strong>{{ team }} Team</strong>
                <div class="menu">
                    <div class="item" v-for="user in users" :data-value="user.id">{{ user.name }}</div>
                </div>
            </div>

            <div class="item" :data-value="currentUserId">Me</div>

        </semantic-form-dropdown>
    </div>
</template>

<script>
    import _ from 'lodash'

    /**
    * A user selector dropdown that outputs a list of teams and each team member within.
    * @example ./croud-supercroud-selector.md
    */

    export default {
        name: 'croud-supercroud-selector',

        props: {
            /**
            * Pass the current users ID
            */
            currentUserId: {
                type: [Number, String],
                default() {
                    return ''
                },
            },

            /**
            * Pass an array of teams
            */
            teamsArray: {
                type: Array,
                default() {
                    return []
                },
            },

            /**
            * Pass any additional users to show in the dropdown. ie SC's with overdue tasks
            */
            options: {
                type: Array,
                default() {
                    return []
                },
            },

            /**
            * The current value of the dropdown
            */
            value: {
                type: [String, Array],
                default() {
                    return []
                },
            },

            /**
            * Defines the placeholder for the dropdown
            */
            dropdownPlaceholder: {
                type: String,
                default() {
                    return 'Select User'
                },
            },

            /**
            * Semantic dropdown settings
            */
            dropdownSettings: {
                type: Object,
                default() {
                    return {
                        useLabels: false,
                        allowCategorySelection: true,
                        onAdd: (val) => {
                            if (isNaN(val)) {
                                const existing = this.value.split(',').map(e => Number(e))
                                this.$nextTick(() => {
                                    this.teams[val].filter(u => existing.indexOf(u.id) === -1).forEach((u) => {
                                        $(this.$refs.supercroudSelector.$el).dropdown('set selected', u.id)
                                    })
                                    $(this.$refs.supercroudSelector.$el).dropdown('remove selected', val)
                                })
                            }
                        },

                        onRemove: (val) => {
                            if (isNaN(val)) {
                                const existing = this.value.split(',').map(e => Number(e))
                                this.$nextTick(() => {
                                    existing.filter(e => this.teams[val].map(u => u.id).indexOf(e) !== -1).forEach((e) => {
                                        $(this.$refs.supercroudSelector.$el).dropdown('remove selected', e)
                                    })
                                })
                            }
                        },
                    }
                },
            },
        },

        data() {
            return {
                users: [],
            }
        },

        computed: {
            model: {
                get() {
                    return this.value
                },

                set(val) {
                    this.$emit('input', val)
                },
            },

            summary() {
                const split = typeof this.model === 'object' ? this.model : this.model.split(',')
                return split.map((user) => {
                    if (this.currentUserId === user) {
                        return 'Me'
                    }

                    const opt = this.users.find(option => option.id === Number(user))
                    return opt ? opt.name : user
                }).join(', ')
            },

            filteredOptions() {
                let teamOptions = []
                Object.keys(this.teams).forEach((key) => {
                    teamOptions = _.unionWith(teamOptions, this.teams[key].map(u => u.id), _.isEqual)
                })
                return this.options.filter(option => option.id !== Number(this.currentUserId)).filter(option => teamOptions.indexOf(option.id) === -1)
            },

            teams() {
                const teams = {}
                this.teamsArray.forEach((team) => {
                    teams[team.name] = team.users.filter(user => user.id !== Number(this.currentUserId))

                    this.users = _.unionWith(this.users, team.users, _.isEqual)
                })
                return teams
            },
        },

        watch: {
            options() {
                this.users = _.unionWith(this.users, this.options, _.isEqual)
            },
        },

        mounted() {
            this.users = _.unionWith(this.users, this.options, _.isEqual)
        },
    }
</script>

<style scoped lang="scss">
    @import '../../../resources/sass/variables/_all.scss';

    .dropdown-wrapper /deep/ .team-aware.ui.selection.dropdown.multiple {
        z-index: $croud-layer-8;

        .menu {
            overflow-x: visible;
            overflow-y: visible;
            max-height: initial;
        }
    }

</style>
