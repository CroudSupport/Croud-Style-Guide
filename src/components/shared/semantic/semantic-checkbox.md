### semantic-checkbox
For our checkboxes, we use the semantic-checkbox component from the [croud-vue-semantic](https://github.com/CroudSupport/vue-semantic) library. Below are a few examples of semantic-checkbox in action. Click [here](http://croudsupport.github.io/vue-semantic/#semantic-checkbox) for full component docs.

### Basic usage

    <semantic-checkbox v-model="checkbox" label="🍕 Pizza" />

### Styling
Pass through the **type** prop to change the style of checkbo

    <div class="ui form">
        <div class="ui field">
            <semantic-checkbox v-model="checkbox" label="Checkbox" />
        </div>
        <div class="ui field">
            <semantic-checkbox v-model="checkbox" label="Slider" type="slider"/>
        </div>
        <div class="ui field">
            <semantic-checkbox v-model="checkbox" label="Toggle" type="toggle" />
        </div>
    </div>

### Disabled

    <semantic-checkbox label="🍕 Pizza" type="toggle" :disabled="true"/>
