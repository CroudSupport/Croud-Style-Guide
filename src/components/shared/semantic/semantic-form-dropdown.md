### semantic-form-dropdown
For our form dropdowns, we use the semantic-form-dropdown component from the [croud-vue-semantic](https://github.com/croudtech/vue-semantic) library. Below are a few examples of semantic-form-dropdown in action. Click [here](http://croudtech.github.io/vue-semantic/#semantic-form-dropdown) for full component docs.

### Basic usage

    <semantic-form-dropdown v-model="dropdownVal" :options="[{id: 1, name: 'Foo'},{id: 2, name: 'Bar'}]"></semantic-form-dropdown>

### Styling
You can use the props to change the styling of the dropdown

    <div class="ui form">
        <div class="field">
            <label>Fluid</label>
            <semantic-form-dropdown fluid v-model="dropdownVal" :options="[{id: 1, name: 'Foo'},{id: 2, name: 'Bar'}]" />
        </div>

        <div class="three fields">
            <div class="field">
                <label>Search</label>
                <semantic-form-dropdown search v-model="dropdownVal" :options="[{id: 1, name: 'Foo'},{id: 2, name: 'Bar'}]" />
            </div>
            <div class="field">
                <label>Multiple</label>
                <semantic-form-dropdown multiple v-model="dropdownVal" :options="[{id: 1, name: 'Foo'},{id: 2, name: 'Bar'}]" />
            </div>
            <div class="field">
                <label>Disabled</label>
                <semantic-form-dropdown :disabled="true" v-model="dropdownVal" :options="[{id: 1, name: 'Foo'},{id: 2, name: 'Bar'}]" />
            </div>
        </div>

        <div class="inline field">
            <label>Inline</label>
            <semantic-form-dropdown :inline="true" v-model="dropdownVal" :options="[{id: 1, name: 'Foo'},{id: 2, name: 'Bar'}]" />
        </div>
    </div>

### Using slots
The **default** slot allows you to add custom options to the top of the dropdown

    <semantic-form-dropdown v-model="dropdownVal" :options="[{id: 1, name: 'Foo'},{id: 2, name: 'Bar'}]">
        <div class="item" v-data="3">
            <b>Baz</b>
        </div>
    </semantic-form-dropdown>

### Handle Events
This component emits the **dropdown-selected** event when an item is selected.

    <semantic-form-dropdown :value="dropdownVal" :options="[{id: 1, name: 'Foo'},{id: 2, name: 'Bar'}]" @dropdown-selected="$toasted.success('TaDa!!!')"></semantic-form-dropdown>
