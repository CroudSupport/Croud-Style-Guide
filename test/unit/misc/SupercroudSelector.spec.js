import Vue from 'vue'
import VueSemantic from 'croud-vue-semantic'
import SupercroudSelector from '../../../src/components/shared/misc/SupercroudSelector'
import '../../../semantic/dist/semantic'

Vue.use(VueSemantic)

const Constructor = Vue.extend(SupercroudSelector)

const vm = new Constructor({
    propsData: {
        teamsArray: [
            { name: 'Supercroud',
                users: [
                    { id: 1, name: 'Joe Blogs' },
                    { id: 2, name: 'John Doe' },
                    { id: 3, name: 'Jason Bourne' },
                ],
            },
        ],
        options: [
            { id: 4, name: 'Jack Overdue' },
        ],
        currentUserId: 1,
    },
}).$mount()

describe('Supercroud selector', () => {
    it('should match the snapshot', () => {
        expect(vm.$el).toMatchSnapshot()
    })

    it('shouldn\'t duplicate current user in team', () => {
        expect(vm.teams.Supercroud.length).toBe(2)
        expect(vm.teams.Supercroud.map(u => u.id)).not.toContain(vm.currentUserId)
    })

    it('should add non current users to filtered options', () => {
        expect(vm.filteredOptions.map(u => u.id)).toContain(vm.options[0].id)
    })
})
