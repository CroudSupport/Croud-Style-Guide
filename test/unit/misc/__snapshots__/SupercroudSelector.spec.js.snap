// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Supercroud selector should match the snapshot 1`] = `
<div
  class="dropdown-wrapper"
>
  <div
    class="team-aware ui selection dropdown multiple"
    data-tooltip=""
    data-inverted="true"
    tabindex="0"
  >
    <input
      type="hidden"
      name=""
    />
    <div
      class="default text"
    >
      Select User
    </div>
    <i
      class="dropdown icon"
    />
    <div
      class="menu dropdown_menu"
      tabindex="-1"
    >
      <div
        data-value="Supercroud"
        class="header item"
      >
        <i
          class="dropdown icon"
        />
        <strong>
          Supercroud Team
        </strong>
        <div
          class="menu"
        >
          <div
            data-value="2"
            class="item"
          >
            John Doe
          </div>
          <div
            data-value="3"
            class="item"
          >
            Jason Bourne
          </div>
        </div>
      </div>
      <div
        data-value="1"
        class="item"
      >
        Me
      </div>
      <div
        data-value="4"
        data-text="Jack Overdue"
        class="item"
      >
        Jack Overdue
      </div>
    </div>
  </div>
</div>
`;
